File: backend/app/routers/v1/twofa.py
from __future__ import annotations

from datetime import datetime, timedelta

from fastapi import APIRouter, HTTPException

from app.core.config import get_settings
from app.core.db import get_db
from app.core.security import UTC, gen_otp, mask_phone, otp_hmac
from app.schemas.twofa import Send2FASmsOtpReq, Verify2FASmsOtpReq
from app.services.strip import strip_private_fields
from app.utils.sms_2factor import send_otp_sms_2factor
from app.core.security import create_token


router = APIRouter(tags=["2fa"])


@router.post("/2fa/sms/send-otp")
async def send_2fa_sms_otp(req: Send2FASmsOtpReq):
    db = get_db()
    ch = await db.sms_otp_challenges.find_one({"id": req.challenge_id, "used": False})
    if not ch:
        raise HTTPException(404, "No active 2FA challenge. Please login again.")
    if datetime.now(UTC) > ch["expires_at"].replace(tzinfo=UTC):
        raise HTTPException(400, "Challenge expired. Please login again.")

    otp = gen_otp()
    await db.sms_otp_challenges.update_one(
        {"_id": ch["_id"]},
        {
            "$set": {
                "otp_hash": otp_hmac(ch["id"], ch["phone"], otp),
                "expires_at": datetime.now(UTC) + timedelta(minutes=10),
                "attempts": 0,
                "created_at": datetime.now(UTC),
            }
        },
    )

    sms_sent = await send_otp_sms_2factor(ch["phone"], otp)
    resp = {"success": True, "challenge_id": ch["id"], "masked_phone": mask_phone(ch["phone"]), "sms_sent": sms_sent}
    settings = get_settings()
    if not settings.twofactor_api_key:
        resp["otp_dev"] = otp
    return resp


@router.post("/2fa/sms/verify-otp")
async def verify_2fa_sms_otp(req: Verify2FASmsOtpReq):
    db = get_db()
    ch = await db.sms_otp_challenges.find_one({"id": req.challenge_id, "used": False})
    if not ch:
        raise HTTPException(400, "No active 2FA challenge. Please login again.")
    if datetime.now(UTC) > ch["expires_at"].replace(tzinfo=UTC):
        raise HTTPException(400, "OTP has expired. Please login again.")
    if ch.get("attempts", 0) >= 5:
        raise HTTPException(429, "Too many attempts. Please login again.")

    expected = ch.get("otp_hash", "")
    got = otp_hmac(ch["id"], ch["phone"], req.otp)
    import hmac as _hmac

    if not _hmac.compare_digest(expected, got):
        await db.sms_otp_challenges.update_one({"_id": ch["_id"]}, {"$inc": {"attempts": 1}})
        raise HTTPException(400, "Invalid OTP. Please try again.")

    await db.sms_otp_challenges.update_one({"_id": ch["_id"]}, {"$set": {"used": True}})

    role = ch["role"]
    user_id = ch["user_id"]
    token = create_token(user_id, role)

    user_data = None
    if role == "User":
        hh = await db.households.find_one({"id": user_id})
        if hh:
            pkg = await db.packages.find_one({"id": hh.get("package_id", "")})
            comm = await db.communities.find_one({"id": hh.get("community_id", "")})
            hh["package"] = strip_private_fields(pkg) if pkg else None
            hh["community"] = strip_private_fields(comm) if comm else None
            user_data = strip_private_fields(hh)
    elif role == "Trainer":
        t = await db.trainers.find_one({"id": user_id})
        if t:
            user_data = strip_private_fields(t)
    elif role == "Manager":
        c = await db.communities.find_one({"id": user_id})
        if c:
            user_data = strip_private_fields(c)
    elif role == "Admin":
        a = await db.admin_users.find_one({"id": user_id})
        if a:
            user_data = strip_private_fields(a)

    return {"token": token, "role": role, "user_id": user_id, "user_data": user_data}

File: backend/app/routes/v1/twofa.js
import express from 'express';
import crypto from 'crypto';
import { settings } from '../../core/config.js';
import { genOtp, maskPhone, otpHmac, createToken } from '../../core/security.js';
import { asyncHandler } from '../../utils/asyncHandler.js';
import SmsOtpChallenge from '../../models/SmsOtpChallenge.js';
import Household from '../../models/Household.js';
import Trainer from '../../models/Trainer.js';
import Community from '../../models/Community.js';
import Admin from '../../models/Admin.js';
import Package from '../../models/Package.js';
import { sendOtpSms2Factor } from '../../utils/sms.js';
import { stripPrivateFields } from '../../utils/helpers.js';

const router = express.Router();

router.post('/2fa/sms/send-otp', asyncHandler(async (req, res) => {
    const { challenge_id } = req.body;
    const ch = await SmsOtpChallenge.findOne({ id: challenge_id, used: false });
    if (!ch) {
        return res.status(404).json({ message: 'No active 2FA challenge. Please login again.', detail: 'No active 2FA challenge. Please login again.' });
    }
    if (new Date() > ch.expiresAt) {
        return res.status(400).json({ message: 'Challenge expired. Please login again.', detail: 'Challenge expired. Please login again.' });
    }

    const otp = genOtp();
    ch.otpHash = otpHmac(ch.id, ch.phone, otp);
    ch.expiresAt = new Date(Date.now() + 10 * 60 * 1000);
    ch.attempts = 0;
    ch.createdAt = new Date();
    await ch.save();

    const smsSent = await sendOtpSms2Factor(ch.phone, otp);
    const resp = {
        success: true,
        challenge_id: ch.id,
        masked_phone: maskPhone(ch.phone),
        sms_sent: smsSent,
    };

    if (!settings.twofactorApiKey) {
        resp.otp_dev = otp;
    }
    res.json(resp);
}));

router.post('/2fa/sms/verify-otp', asyncHandler(async (req, res) => {
    const { challenge_id, otp } = req.body;
    const ch = await SmsOtpChallenge.findOne({ id: challenge_id, used: false });
    if (!ch) {
        return res.status(400).json({ message: 'No active 2FA challenge. Please login again.', detail: 'No active 2FA challenge. Please login again.' });
    }
    if (new Date() > ch.expiresAt) {
        return res.status(400).json({ message: 'OTP has expired. Please login again.', detail: 'OTP has expired. Please login again.' });
    }
    if (ch.attempts >= 5) {
        return res.status(429).json({ message: 'Too many attempts. Please login again.', detail: 'Too many attempts. Please login again.' });
    }

    const expected = ch.otpHash;
    const got = otpHmac(ch.id, ch.phone, otp);

    if (!crypto.timingSafeEqual(Buffer.from(expected), Buffer.from(got))) {
        ch.attempts += 1;
        await ch.save();
        return res.status(400).json({ message: 'Invalid OTP. Please try again.', detail: 'Invalid OTP. Please try again.' });
    }

    ch.used = true;
    await ch.save();

    const { role, userId } = ch;
    const token = createToken(userId, role);

    let userData = null;
    if (role === 'User') {
        const hh = await Household.findOne({ id: userId });
        if (hh) {
            const pkg = await Package.findOne({ id: hh.package_id });
            const comm = await Community.findOne({ id: hh.community_id });
            const hhObj = hh.toObject();
            hhObj.package = pkg ? stripPrivateFields(pkg) : null;
            hhObj.community = comm ? stripPrivateFields(comm) : null;
            userData = stripPrivateFields(hhObj);
        }
    } else if (role === 'Trainer') {
        const t = await Trainer.findOne({ id: userId });
        if (t) userData = stripPrivateFields(t);
    } else if (role === 'Manager') {
        const c = await Community.findOne({ id: userId });
        if (c) userData = stripPrivateFields(c);
    } else if (role === 'Admin') {
        const a = await Admin.findOne({ id: userId });
        if (a) userData = stripPrivateFields(a);
    }

    res.json({ token, role, user_id: userId, user_data: userData });
}));

export default router;
